theRAPTLab/gsgo
# Support arbitrary screen and project aspect ratios in CharControl3

## Problems

There were two problems:
1. If the CharControl3 *screen* was narrower than the current project setting (e.g. where the blank black sidebars next to the simulation were not visible), then the drag area would be wrong and you would lose the dragger cursor.
2. If the *project settings* used a different aspect ratio (e.g. 800 wide x 400 high), then the drag area would be wrong and you would lose the dragger cursor.

## The Fix

Redid the algorithm for sizing the drag container to NOT assume a square aspect ratio.  Load the current project settings to use the project-defined simulation size when calculating the drag container.

## Other Fixes

* Simplified layout layering to improve performance -- there can be performance issues when the simulation was being drawn on top of the dragger layer.
* Cleaned up react references


## To Test
1. Resize CharController to be a very narrow window taller than it is wide -- the dragger cursor should still be visible.
2. Change Project Settings to `right: 800`, `left: -800`.  Save.  Reload.  The dragger cursor should still be visible.
3. Quickly drag the cursor back and forth to make sure the dragger can keep up with the character.


## Caveats

Project Settings actually supports arbitrary top/left/bottom/right values with the goal of supporting asymmetrical stage areas (e.g. you can have the stage offset to the right).  The drag container does not currently support this.  There may also be issues with the Viewer and other displays when using asymmetrical dimensions.
